### ISSUE TITLE
Can convert Nill into String when the Image uri is wrong (no extension)

### ISSUE THREAD

**Issue by cmoulliard** [asciidoctor/asciidoctor#1455]
The existing code of the function generate_data_uri takes the assumption that the uri string is correct and contains the file extension 

```
https://github.com/asciidoctor/asciidoctor/blob/master/lib/asciidoctor/abstract_node.rb#L327-L330
```

this is why we get this error :  

```
abstract_node.rb:327:in `+': can't convert nil into String (TypeError)
```

if the image uri is wrong ( example --> image:messaging.[])

**Comment by cmoulliard** [comment]
Should be great to report not only the image name but also the asciidoc file where the exception occurred :  

````
    if ext.nil?
      warn "asciidoctor: WARNING: image extension to embed is empty for the image name '#{target_image}' part of the document '#{@document.name}'"
      ext = ""
    end
````

**Comment by mojavelinux** [comment]
Currently, Asciidoctor requires that image paths have extensions to indicate type. I'd rather avoid getting into attempting to read the image data and parse out the type, but that doesn't mean we won't do it.

It's not appropriate to embed more metadata directly into the warning message. That's better handled as a cross cutting concern, which is why the focus should be on resolving #44 (which is scheduled).

**Comment by mojavelinux** [comment]
For now, I'd like to approach this by emphasizing in the docs that a file extension is required if you are going to embed images (linking to them works fine without the extension).

**Comment by cmoulliard** [comment]
Adding  a remark to the doc is not enough as some users could be interested to get the warning message containing the erroneous file and where (= which adoc file) the error happened

**Comment by mojavelinux** [comment]
I wasn't suggesting the remark is enough, but rather that the error message can be properly decorated if we use a logger rather than (somewhat) arbitrarily adding things to the error message. I think the message needs to be about what happened, not about what the context is at the origin of error. So we agree, just perhaps viewing the approach to a solution differently.

**Comment by mojavelinux** [comment]
This was fixed 1.5.0.
